plotter/heat.go

```diff
@@ -0,0 +1,210 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package plotter
+
+import (
+	"image/color"
+	"math"
+
+	"github.com/gonum/plot"
+	"github.com/gonum/plot/palette"
+	"github.com/gonum/plot/vg"
+	"github.com/gonum/plot/vg/draw"
+)
+
+// GridXYZ describes three dimensional data where the X and Y
+// coordinates are arranged on a rectangular grid.
+type GridXYZ interface {
+	// Dims returns the dimensions of the grid.
+	Dims() (c, r int)
```

Comment:
My version should give:

    1-2x-3-4-5y-6-7-8y-9-10x-11
    	5 6 7 8
    	2 3 4 5 9 10
    	1 2 11
    
    1-2xy-3-4-5-6-7-8y-9-10x-11
    	2 3 4 5 6 7 8
    	2 9 10
    	1 2 11
    
    1-2y-3-4-5x-6-7-8y-9-10x-11
    	2 3 4 5 6 7 8
    	1 2 9 10 11

I.e. once it detects a loop, it extracts the loop from the content as if it never happened and ignores it for the rest of processing. This should create good enough paths - but depending on the place where you start iterating you may get different results... i.e. the last case would give different results if the sequence were reversed. Also I'm not sure whether it's worth the effort to fix that... I assume this would mean constructing all possible loops, evaluating based on some metric and start removing them one-by-one.